File: bindings/typescript/src/safe_url.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeUrl - Typed wrapper for URL parsing that cannot crash.
 *
 * Delegates all computation to the JavaScript FFI binding, which calls
 * libproven (Idris 2 + Zig) via Deno.dlopen. No logic is reimplemented here.
 *
 * @module
 */

import { SafeUrl as JsSafeUrl, Url as JsUrl } from '../../javascript/src/safe_url.js';

/** Typed re-export of parsed URL components. */
export interface ParsedUrl {
  scheme: string;
  host: string;
  port: number | null;
  path: string;
  query: string | null;
  fragment: string | null;
}

/** Result type for URL operations. */
export type UrlResult<T> =
  | { readonly ok: true; readonly value: T }
  | { readonly ok: false; readonly error: string };

/**
 * Safe URL operations backed by formally verified Idris 2 code.
 * All methods delegate to the JavaScript FFI wrapper.
 */
export class SafeUrl {
  /**
   * Parse a URL into its components.
   * Delegates to the JavaScript FFI URL parser.
   *
   * @param url - URL string to parse.
   * @returns Parsed URL components or null on failure.
   */
  static parse(url: string): ParsedUrl | null {
    const result = JsSafeUrl.parse(url);
    if (!result || !result.ok) return null;
    return result.value as ParsedUrl;
  }

  /**
   * Check if a URL is valid.
   *
   * @param url - URL string to validate.
   * @returns true if the URL can be parsed.
   */
  static isValid(url: string): boolean {
    return SafeUrl.parse(url) !== null;
  }

  /**
   * Extract the host from a URL.
   *
   * @param url - URL string.
   * @returns The host component, or null on parse failure.
   */
  static getHost(url: string): string | null {
    const parsed = SafeUrl.parse(url);
    return parsed?.host ?? null;
  }

  /**
   * Extract the scheme from a URL.
   *
   * @param url - URL string.
   * @returns The scheme component, or null on parse failure.
   */
  static getScheme(url: string): string | null {
    const parsed = SafeUrl.parse(url);
    return parsed?.scheme ?? null;
  }

  /**
   * Check if a URL uses HTTPS.
   *
   * @param url - URL string.
   * @returns true if the scheme is "https".
   */
  static isHttps(url: string): boolean {
    const scheme = SafeUrl.getScheme(url);
    return scheme?.toLowerCase() === 'https';
  }
}

File: bindings/julia/src/SafeUrl.jl
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: 2025 Hyperpolymath

"""
    SafeUrl

Safe URL parsing and validation operations.
"""
module SafeUrl

export ParsedUrl, parse_url, is_valid_url, get_host, get_path, is_https

"""
    ParsedUrl

Represents the parsed components of a URL.
"""
struct ParsedUrl
    scheme::String
    host::String
    port::Union{Int, Nothing}
    path::String
    query::Union{String, Nothing}
    fragment::Union{String, Nothing}
end

"""
    parse_url(url::AbstractString) -> Union{ParsedUrl, Nothing}

Parse a URL into its components.
"""
function parse_url(url::AbstractString)::Union{ParsedUrl, Nothing}
    # Match scheme
    scheme_match = match(r"^([a-zA-Z][a-zA-Z0-9+.-]*)://", url)
    scheme_match === nothing && return nothing

    scheme = lowercase(scheme_match.captures[1])
    rest = url[length(scheme_match.match)+1:end]

    # Split off fragment
    fragment = nothing
    if contains(rest, "#")
        idx = findfirst('#', rest)
        fragment = rest[idx+1:end]
        rest = rest[1:idx-1]
    end

    # Split off query
    query = nothing
    if contains(rest, "?")
        idx = findfirst('?', rest)
        query = rest[idx+1:end]
        rest = rest[1:idx-1]
    end

    # Split host and path
    path = "/"
    if contains(rest, "/")
        idx = findfirst('/', rest)
        path = rest[idx:end]
        rest = rest[1:idx-1]
    end

    # Parse host and port
    port = nothing
    host = rest
    if contains(rest, ":")
        idx = findfirst(':', rest)
        host = rest[1:idx-1]
        port_str = rest[idx+1:end]
        port = tryparse(Int, port_str)
        port === nothing && return nothing
        (port < 0 || port > 65535) && return nothing
    end

    isempty(host) && return nothing

    ParsedUrl(scheme, host, port, path, query, fragment)
end

"""
    is_valid_url(url::AbstractString) -> Bool

Check if a string is a valid URL.
"""
function is_valid_url(url::AbstractString)::Bool
    parse_url(url) !== nothing
end

"""
    get_host(url::AbstractString) -> Union{String, Nothing}

Extract the host from a URL.
"""
function get_host(url::AbstractString)::Union{String, Nothing}
    parsed = parse_url(url)
    parsed === nothing ? nothing : parsed.host
end

"""
    get_path(url::AbstractString) -> Union{String, Nothing}

Extract the path from a URL.
"""
function get_path(url::AbstractString)::Union{String, Nothing}
    parsed = parse_url(url)
    parsed === nothing ? nothing : parsed.path
end

"""
    is_https(url::AbstractString) -> Bool

Check if a URL uses HTTPS.
"""
function is_https(url::AbstractString)::Bool
    parsed = parse_url(url)
    parsed !== nothing && parsed.scheme == "https"
end

end # module
